scripts/src/java/org/oppia/android/scripts/coverage/CoverageRunner.kt

```diff
@@ -0,0 +1,79 @@
+package org.oppia.android.scripts.coverage
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.async
+import org.oppia.android.scripts.common.BazelClient
+import org.oppia.android.scripts.common.CommandExecutor
+import org.oppia.android.scripts.common.CommandExecutorImpl
+import org.oppia.android.scripts.common.ScriptBackgroundCoroutineDispatcher
+import java.io.File
+import java.util.concurrent.TimeUnit
+
+/**
+ * Class responsible for running coverage analysis asynchronously.
+ *
+ * @param repoRoot the absolute path to the working root directory
+ * @param targetFile Path to the target file to analyze coverage.
+ */
+class CoverageRunner(
+  private val repoRoot: File,
+  private val scriptBgDispatcher: ScriptBackgroundCoroutineDispatcher
+) {
+
+  /**
+   * Runs coverage analysis asynchronously for the Bazel test target.
+   *
+   * @param repoRoot the absolute path to the working root directory
+   * @param scriptBgDispatcher the [ScriptBackgroundCoroutineDispatcher] to be used for running the coverage command
+   * @param bazelTestTarget Bazel test target to analyze coverage.
+   * @return a deferred value that contains the path of the coverage data file [will contain the proto for the coverage data].
+   */
+  suspend fun runWithCoverageAsync(
+    bazelTestTarget: String
+  ): Deferred<String?> {
+    return CoroutineScope(scriptBgDispatcher).async {
+      val coverageData = getCoverage(bazelTestTarget)
+      val data = coverageData
+      parseCoverageDataFile(data)
+    }
+  }
+
+  /**
+   * Runs coverage command for the Bazel test target.
+   *
+   * @param repoRoot the absolute path to the working root directory
+   * @param scriptBgDispatcher the [ScriptBackgroundCoroutineDispatcher] to be used for running the coverage command
+   * @param bazelTestTarget Bazel test target to analyze coverage.
+   * @return a lisf of string that contains the result of the coverage execution.
+   */
+  fun getCoverage(
+    bazelTestTarget: String
+  ): List<String> {
+    val commandExecutor: CommandExecutor = CommandExecutorImpl(
+      scriptBgDispatcher, processTimeout = 5, processTimeoutUnit = TimeUnit.MINUTES
+    )
+    val bazelClient = BazelClient(repoRoot, commandExecutor)
+    val coverageData = bazelClient.runCoverageForTestTarget(bazelTestTarget)
+    return coverageData
+  }
+
+  /**
+   * Parse the coverage command result to extract the path of the coverage data file.
+   *
+   * @param data the result from the execution of the coverage command
+   * @return the extracted path of the coverage data file.
+   */
+  fun parseCoverageDataFile(data: List<String>): String? {
```

Comment:
This & the method above don't need to be public since they're never called outside this class--suggest making them private and/or inlining them.

They also won't need KDocs.

Reply:
**Update:** Revoked those visibility changes [[commit](https://github.com/oppia/oppia-android/pull/5423/commits/6cb1da7056f63c85f1327ba0366594b43428cf6d)] as they were called in `BazelClientTest` testing methods [[BazelClientTest.kt](https://github.com/oppia/oppia-android/pull/5423/files#diff-e50d7ab1968e374641f7298e27dca62ec182a42dc06dac4c47857cd7526de00fR406)].

Reply:
Do they need to be? Generally we don't want to expose helper methods for the purpose of testing. Often we call helpers in tests for one of two reasons:
- We want to test that the helper is doing the right thing. The better practice is to figure out what situations of calling methods lead to that state of the helper method and verify those, instead. I call this process mapping certain program states back to public API behaviors that can be verified in tests.
- We need to use the helper to verify something, or to tweak conditions for a public API method. This is definitionally a "back door" into the program state for the sake of testing, and it's considered a bad practice because it increases the surface area of the public API of the class being tested past what we actually want to expose to non-test callers. Simply put, all public API methods/fields are "fair game" to be called in any place that has access to the Bazel target, so we should only expose things we want to be called outside of test cases.

Reply:
Makes sense with this information and what you explained in the meeting. Thanks, Ben. 
Also, aligning with this, I made all the helper methods private and updated the related tests.